### ISSUE TITLE
ModelvalidationException falsepositive in non-english environment

### ISSUE THREAD

**Issue by andreasblueher** [SubPointSolutions/spmeta2#982]
Hey,
I've run into the following issue. If I use ContentTypeId in AddRemoveContentTypeLinks I get the attached error message. Prior to using .110 I've never had this issue so I guess it has been implemented between .60 and 110.

![a9e488da](https://cloud.githubusercontent.com/assets/12544512/24103958/356aa14c-0d81-11e7-8e42-63e5f324664d.PNG)

I get why the validation rule makes sense in english enviroments, but in non-english setups using the id is the only valid option.

### Brief description
Please provide short description covering two areas:
* expected behaviour
one should be able to remove contenttypelinks from lists
* actual behavior
one has to use "Dokument" or "Aufgabe" instead of BuiltInContentTypeNames to have it working in a german site

### SharePoint API
Which version of SharePoint runtime do you use? 
* CSOM / SSOM / O365?
SSOM SharePoint 2016 On-Premise

### SPMeta2 API
Which version of SPMeta2 do you use? Use the following code snippet for your convenience. It shows all the details on SPMeta2 and SharePoint runtime used.
1.2.110

**Comment by SubPointSupport** [comment]
Chanhe HideContentTypeLinksDefinition to use not Id but Name. Al would be fine in that way.

**Comment by andreasblueher** [comment]
SharePoint Contenttypes in german sitecollection are named "Dokument", "Aufgabe" or "Ereignis". So if I have to use Name for AddRemoveContentTypeLinks I can't use "BuiltInContentTypeIds" anymore and have to use MagicStrings in my code. This code will fail if I want to deploy it on a english site collection.

I just created a new doclib on my dev system and the doc contenttype has id 0x0101 which would allow you to use it. For a new task list I got 0x0108. So for default contenttypes it should work (given the tests I just did) and imho removing default contenttypes should be the default case. Maybe there could be an option "AddRemoveDefaultContentTypeLink" which would allow you to use id in addiation to name.

Given the current status I can't remove contenttypes in multilingual setups.

**Comment by SubPointSupport** [comment]
First of all, check out the following page on Id/Name explanation:
[Content type operations within a list](http://docs.subpointsolutions.com/spmeta2/kb/kb-m2-000003.html)

List scope content type can't be found by ID due to its dynamic nature.

Secondly, is there any reason content type Name can't be used to lookup the content type? If you know the target site, you can always get the locale and hence the name of the content type in the target locale. Not that we can't do that (you can always remove validation from the .PreDeploymentService of the provision service instance) but that we can't lookup the target content type by ID always correctly. That does not really make any sense unless we specify the name of the content type. 

If you can do that via SharePoint API - great, we are happy to integrate that into SPMeta2. However, you would end up with content type ID matching which wouldn't be always correct. If you fine with that, remove validation and go for it. That's fine. if there is a better workaround - we are happy to add into M2.

**Comment by andreasblueher** [comment]
I agree to every point you made, but there is a small special case I'm talking about which doesn't apply to the general case:

When a new list is created using a template like "Task" the content type attached to the list has the id 0x0108. If I would create a custom list and adding the task content type to the list it would get 0x010800GUID as it's id. You're talking about the second case and you're right, you cannot identify the contenttype by 0x0108 because there might be more than one ct based on task. Rather you require the name which has to be unique.

My argument is that every content type which has been by creating the list in the first place does use the simple (not combined) id. I don't want to remove multiple content types and rather remove the one which has been there in the first place, after I added a second one.

One other solution would be to have the translations for all BuiltInContentTypeNames included and choose automatically, but adding the language specific name of the content type to the model doesn't feel like a good solution so the only option would be to do it manually and I would rather not.

**Comment by SubPointSupport** [comment]
>>My argument is that every content type which has been by creating the list in the first place does use the simple (not combined) id. I don't want to remove multiple content types and rather remove the one which has been there in the first place, after I added a second one.

So, that means we would find and remove content type by ID and ONLY fulll ID match rather than "best match" based on string.StartsWith(), right? That would work for out of the box content types, probably. We need to investigate and confirm that.

The problem here is that we had been having a fair share complains on "SPMeta2 does not find content type by ID" or "SPMeta2 finds wrong content types by ID" within a list. That's been a real pain so that at some point we decided to enforce validation and avoid issues before they even happen. 

>>One other solution would be to have the translations for all BuiltInContentTypeNames included and choose automatically, but adding the language specific name of the content type to the model doesn't feel like a good solution so the only option would be to do it manually and I would rather not.

Yes, indeed. Content type name becomes an identity key here. Unfortunately, we are bound to SharePoint API and its limitations. ID does not cover all cases and only Name provides a robust way to work with all content type operations within a list. You are right, that becomes messy with multi-language setup and, most-likely, adding some special property to indicate other content type name might be a way to go.

The fundamental issue is that SPMeta2 does not do a good job for contenttype operations within a list over multi-language setup, right? That includes Add/Remove/Hide operations while finding an actual content type under multi-language setup. So, we probably need to focus on improving that bit.

**Comment by andreasblueher** [comment]
Hey @SubPointSupport,

this time I fully agree with your reply. I'll try to implement something which could be used to remove "Default" contenttypes like a third option next to ContentTypeName and ContentTypeId, maybe a property called DefaultContentTypeId which could be validated if it contains "00" which would indicate that someone is trying to enter a combined ID (which should throw an error). Shouldn't be to complicated (at least I hope so :o)

I completly understand why you enforced the validation, but I also agree that there is place for improvement regarding multi-language setups. BuiltIn Lists are basically the same problem, where you can't use CustomUrl property because the url will be different in a primarely german setup.

On the other side I'm not sure how a good solution for multilanguage setups could look like, because integrating all translations into your solutions would be a project on it's own.

Just a quick thought, maybe for contenttypes for non-english sites you could retrieve the name of the default contenttypes by getting them by id and there you would have the translation. (not sure if you understand what I mean)

**Comment by SubPointSupport** [comment]
Well, first of all, that validation can be disabled. 

Next, agree, we can change lookup method of the content type within list limiting it to ID match rather than  best match (partial match). That would be more than enough to handle default content type Hide/Remove operations, right?

**Comment by andreasblueher** [comment]
Yes this would be enough and could you link me the code on how to remove this validation? I tried this code, but it didn't work.

```
provisionService.PreDeploymentServices.Remove(new DefaultContentTypeLinkValidationService());
```


If you decide to replace the logic to identify content types not using best match anymore you can do so. The reason why I though about adding another property was to not change existing methods, because people could have disabled validation for their model and still be using it.

**Comment by SubPointSupport** [comment]
@andreasblueher, FYI

* .AddContentTypeLink() should work with content type ID #1016
* .RemoveContentTypeLink() should work with content type ID #1020
* .HideContentTypeLink() should work with content type ID #1021
* .AddUniqueContentTypeOrder() should work with content type ID #1022

**Comment by andreasblueher** [comment]
Great! This means 1.2.130 will contain this change? When will it be released?

**Comment by SubPointSupport** [comment]
Correct, coming in 1.2.130. Finishing up testing, ETA is early next week.

**Comment by andreasblueher** [comment]
Great, this means I go update from 1.2.60 to 1.2.130 to have incremental updates and renewed logic for handling content types.

In 1.2.60 we could still use ContentTypeId so not having an equivalent option we couldn't upgrade yet. Loooking forward to do it next week.

**Comment by SubPointSupport** [comment]
> In 1.2.60 we could still use ContentTypeId so not having an equivalent option we couldn't upgrade yet. 
Wow! Frankly saying, we didn't know and didn't expect that this issue is so severe that you didn't upgrade ever since. That needs to change so that first of all, you would be able to upgrade without any fear and secondly, you would be happy. Please let us know if there is anything else that is really critical for you so that we can see what can be done or impoved.

**Comment by andreasblueher** [comment]
That sounded most likely worse than I intended it to be. Given the multilingual (3 different languages) environment using id was our only option from the beginning. It wasn't until 1.2.110 or so when I realized that you added the new validations, but we had other (bigger) issues in the project, so that was no real concern. I like to take the "don't fix it if it ain't broken" route rather than installing every update, but incremental updates are to sweet to not have them.

I'm more than happy having and using SPMeta2 in my projects and every member of my team would confirm that :D. I'm planning on looking into MetaPack next week and maybe even start using this.
